File: Baitap3/Hieu/merge_sort.py
import datetime

def __validate_(arr):
    if (type(arr) is not list):
        return False
    for x in arr:
        if (type(x) is not int):
            return False
    return True

def __merge_(arr, fr, middle, to):
    cache = arr[fr:to]
    idLeft = fr; idRight = middle; iterPos = fr
    while (idLeft < middle and idRight < to):
        if (cache[idLeft - fr] < cache[idRight - fr]):
            arr[iterPos] = cache[idLeft - fr]
            idLeft += 1
        else:
            arr[iterPos] = cache[idRight - fr]
            idRight += 1
        iterPos += 1

    while (idLeft < middle):
        arr[iterPos] = cache[idLeft - fr]
        iterPos += 1
        idLeft += 1

    while (idRight < to):
        arr[iterPos] = cache[idRight - fr]
        iterPos += 1
        idRight += 1

def __sort_recursive_(arr, fr, to):
    length = to - fr
    if (length <= 1):
        return
    middle = fr + (length >> 1)
    __sort_recursive_(arr, fr, middle)
    __sort_recursive_(arr, middle, to)
    __merge_(arr, fr, middle, to)

def sort(arr):
    if (not __validate_(arr)):
        raise ValueError("arr is not a list of int!")
    __sort_recursive_(arr, 0, len(arr))

def main():
    inputFile = open("test.txt", "r")
    n = int(inputFile.readline())
    array = list(map(int, inputFile.readline().split()))

    start_time = datetime.datetime.now()
    sort(array)
    end_time = datetime.datetime.now()
    print(end_time - start_time)
    
if __name__ == '__main__':
    main()

File: Baitap3/Hieu/MergeSort.java
import java.io.File;
import java.time.Instant;
import java.util.Scanner;

public class MergeSort {
    public static void main(String[] args) {
        File inputFile = new File("test.txt");
        Scanner scanner;
        try {
            scanner = new Scanner(inputFile);    
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }
        int n = scanner.nextInt();
        int[] arr = new int[n];
        for(int i = 0; i < n; i ++)
            arr[i] = scanner.nextInt();
        
        long startTime = Instant.now().toEpochMilli();
        sort(arr);
        long endTime = Instant.now().toEpochMilli();
        long runTime = endTime - startTime;
        System.out.println(runTime);
        scanner.close();
    }

    private static int[] cache;

    public static void sort(int[] arr) {
        if (arr == null)
            return;
        cache = new int[arr.length];
        sortRecursive(arr, 0, arr.length);
    }

    private static void sortRecursive(int[] arr, int from, int to) {
        int length = to - from;
        if (length <= 1)
            return;
        int middle = from + (length >> 1);
        sortRecursive(arr, from, middle);
        sortRecursive(arr, middle, to);
        merge(arr, from, middle, to);
    }

    public static void merge(int[] arr, int from, int middle, int to) {
        for(int i = from; i < to; i ++)
            cache[i] = arr[i];
        int idLeft = from, idRight = middle, iterPos = from;
        while(idLeft < middle && idRight < to) {
            if (cache[idLeft] < cache[idRight])
                arr[iterPos ++] = cache[idLeft ++];
            else
                arr[iterPos ++] = cache[idRight ++];
        }
        while(idLeft < middle)
            arr[iterPos ++] = cache[idLeft ++];
        while(idRight < to)
            arr[iterPos ++] = cache[idRight ++];
    }
}